precompiles/staking/staking.go

```diff
@@ -228,6 +240,41 @@ func (c *Contract) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) ([]byt
 			return res, err
 		}
 		return res, nil
+	case GetRewardsMethodName:
+		var res []byte
+		execErr := stateDB.ExecuteNativeAction(contract.Address(), nil, func(ctx sdk.Context) error {
+			res, err = c.getRewards(ctx, method, args)
+			return err
+		})
+		if execErr != nil {
+			return nil, err
```

Comment:
As far as I know execErr would be the error returned from `statedb`, and the error we'd want to know is the one returned from the function being called, that is `err`.

Reply:
I see what you mean now. I reviewed the ExecuteNativeAction code and it's the following:

```go
func (s *StateDB) ExecuteNativeAction(contract common.Address, converter EventConverter, action func(ctx sdk.Context) error) error {
	snapshot := s.cloneNativeState()
	eventManager := sdk.NewEventManager()

	if err := action(s.cacheCtx.WithEventManager(eventManager)); err != nil {
		s.restoreNativeState(snapshot)
		return err
	}

	events := eventManager.Events()
	s.emitNativeEvents(contract, converter, events)
	s.nativeEvents = s.nativeEvents.AppendEvents(events)
	s.journal.append(nativeChange{snapshot: snapshot, events: len(events)})
	return nil
}
```

The only way for it to return an error is if the `action` (the precompiled func) returns an error, otherwise returns nil. 
So it would be the same returning `execErr` or `err` from the precompile code:

```go
	case ClaimRewardsMethodName:
		var res []byte
		execErr := stateDB.ExecuteNativeAction(contract.Address(), nil, func(ctx sdk.Context) error {
			res, err = c.claimRewards(ctx, evm, contract, method, args)
			return err
		})
		if execErr != nil {
			res, errPack := method.Outputs.Pack(false)
			if errPack != nil {
				return nil, errPack
			}

			return res, err
		}
		return res, nil
```

The only distinction I make is when the error is returned from pack, which would mean the ABI is not correctly loaded.